unison-share-api/src/Unison/Server/Backend.hs

```diff
@@ -1127,23 +1129,21 @@ scopedNamesForBranchHash ::
   Backend m (Names, PPED.PrettyPrintEnvDecl)
 scopedNamesForBranchHash codebase mbh path = do
   shouldUseNamesIndex <- asks useNamesIndex
+  (rootBranchHash, rootCausalHash) <- case mbh of
+    Just cb -> pure (V2Causal.valueHash cb, V2Causal.causalHash cb)
+    Nothing -> lift $ do
+      cb <- Codebase.runTransaction codebase Operations.expectRootCausal
+      pure (V2Causal.valueHash cb, V2Causal.causalHash cb)
+  haveNameLookupForRoot <- lift $ Codebase.runTransaction codebase (Ops.checkBranchHashNameLookupExists rootBranchHash)
   hashLen <- lift $ Codebase.runTransaction codebase Codebase.hashLength
-  (parseNames, localNames) <- case mbh of
-    Nothing
-      | shouldUseNamesIndex -> do
-          lift $ Codebase.runTransaction codebase indexNames
-      | otherwise -> do
-          rootBranch <- lift $ Codebase.getRootBranch codebase
-          let (parseNames, _prettyNames, localNames) = namesForBranch rootBranch (AllNames path)
-          pure (parseNames, localNames)
-    Just rootCausal -> do
-      let ch = V2Causal.causalHash rootCausal
-      rootHash <- lift $ Codebase.runTransaction codebase Operations.expectRootCausalHash
-      if (ch == rootHash) && shouldUseNamesIndex
-        then lift $ Codebase.runTransaction codebase indexNames
-        else do
-          (parseNames, _pretty, localNames) <- flip namesForBranch (AllNames path) <$> resolveCausalHash (Just ch) codebase
-          pure (parseNames, localNames)
+  (parseNames, localNames) <-
```

Comment:
We simplify the logic here a little;

Previously we would check if the requested branch hash was equal to the current  root hash, because if it was we could use the root names lookup;

Now that we can have multiple indexes, we can just always check if we have an index  for a given hash, and bail if we don't (Enlil guarantees we'll always have an index for every branch HEAD and user codebase root)